JacksonDunstanNativeCollections/NativePerJobThreadIntPtrs.cs

```diff
@@ -0,0 +1,434 @@
+	[DebuggerDisplay("Value = NativeArray")]
+	[StructLayout(LayoutKind.Sequential)]
+	public unsafe struct NativePerJobThreadIntPtrs : IDisposable
+	{
+		/// <summary>
+		/// An atomic write-only version of the object suitable for use in a
+		/// ParallelFor job
+		/// </summary>
+		[NativeContainer]
+		[NativeContainerIsAtomicWriteOnly]
+		public struct Parallel {
+
+			/// <summary>
+			/// The number of int ptrs.
+			/// </summary>
+			public int Length;
+
+			/// <summary>
+			/// Pointer to the value in native memory
+			/// </summary>
+			[NativeDisableUnsafePtrRestriction]
+			internal int* m_Buffer;
+
+			/// <summary>
+			/// Thread index of the job using this object. This is set by Unity
+			/// and must have this exact name and type.
+			/// </summary>
+			[NativeSetThreadIndex]
+			internal int m_ThreadIndex;
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+			/// <summary>
+			/// A handle to information about what operations can be safely
+			/// performed on the object at any given time.
+			/// </summary>
+			internal AtomicSafetyHandle m_Safety;
+
+			/// <summary>
+			/// Create a parallel version of the object
+			/// </summary>
+			/// 
+			/// <param name="value">
+			/// Pointer to the value
+			/// </param>
+			/// 
+			/// <param name="safety">
+			/// Atomic safety handle for the object
+			/// </param>
+			internal Parallel(int length, int* value, AtomicSafetyHandle safety)
+			{
+				Length = length;
+				m_Buffer = value;
+				m_ThreadIndex = 0;
+				m_Safety = safety;
+			}
+#else
+			/// <summary>
+			/// Create a parallel version of the object
+			/// </summary>
+			/// 
+			/// <param name="value">
+			/// Pointer to the value
+			/// </param>
+			internal Parallel(int* value)
+			{
+				m_Buffer = value;
+				m_ThreadIndex = 0;
+			}
+#endif
+
+			/// <summary>
+			/// Increment the stored value
+			/// </summary>
+			/// 
+			/// <returns>
+			/// This object
+			/// </returns>
+			[WriteAccessRequired]
+			public void Increment(int index)
+			{
+				RequireWriteAccess();
+				m_Buffer[IntsPerCacheLine * m_ThreadIndex + index]++;
+			}
+
+			/// <summary>
+			/// Decrement the stored value
+			/// </summary>
+			/// 
+			/// <returns>
+			/// This object
+			/// </returns>
+			[WriteAccessRequired]
+			public void Decrement(int index)
+			{
+				RequireWriteAccess();
+				m_Buffer[IntsPerCacheLine * m_ThreadIndex + index]--;
+			}
+
+			/// <summary>
+			/// Add to the stored value
+			/// </summary>
+			/// 
+			/// <param name="value">
+			/// Value to add. Use negative values for subtraction.
+			/// </param>
+			/// 
+			/// <returns>
+			/// This object
+			/// </returns>
+			[WriteAccessRequired]
+			public void Add(int index, int value)
+			{
+				RequireWriteAccess();
+				m_Buffer[IntsPerCacheLine * m_ThreadIndex + index] += value;
+			}
+
+			/// <summary>
+			/// Throw an exception if the object isn't writable
+			/// </summary>
+			[Conditional("ENABLE_UNITY_COLLECTIONS_CHECKS")]
+			[BurstDiscard]
```

Comment:
This shouldn't have `[BurstDiscard]` because we want the checks when Burst is enabled with safety checks.